test/distributed/pipelining/test_composability.py

```diff
@@ -0,0 +1,230 @@
+    requires_nccl,
+)
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+    skip_but_pass_in_sandcastle_if,
+)
+
+
+class ComposabilityTest(MultiProcContinousTest):
+    @classmethod
+    def backend_str(cls) -> str:
+        # Testing with NCCL backend
+        return "nccl"
+
+    @classmethod
+    def setUpClass(cls):
+        """
+        Class-scope test fixture. Run once for entire test class, before any test starts.
+        Set up the device.
+        """
+        super().setUpClass()
+        dev_id = cls.rank % torch.cuda.device_count()
+        cls.device = torch.device(f"cuda:{dev_id}")
+        # TODO: investigate why this is needed to prevent multiple NCCL ranks from hitting the same device
+        torch.cuda.set_device(cls.device)
+
+    @requires_nccl()
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "Test requires 4+ GPUs")
+    @parametrize("dp_type", ["DDP", "FSDP"])
+    @parametrize("ScheduleClass", [ScheduleGPipe, Schedule1F1B])
+    def test_manual_with_data_parallel(self, dp_type, ScheduleClass):
+        device_mesh = init_device_mesh(
+            "cuda", mesh_shape=(2, 2), mesh_dim_names=("dp", "pp")
+        )
+        pp_group = device_mesh["pp"].get_group()
+        dp_mesh = device_mesh["dp"]
+
+        # create "entire model"
+        total_layers = 8
+        dim = 10
+        full_model = nn.ModuleList([MLPModule(dim) for _ in range(total_layers)])
+        ref_model = nn.Sequential(*copy.deepcopy(full_model))
+        ref_model.to(self.device)
+
+        # Prepare inputs
+        num_microbatches = 8
+        inputs = [
+            torch.rand((num_microbatches, dim), device=self.device)
+            for _ in range(dp_mesh.size())
+        ]
+        input = inputs[dp_mesh.get_local_rank()]
+        input_mb = [[input[i].reshape((1, dim))] for i in range(num_microbatches)]
+
+        # dummy loss needed just to force backwards to run in schedule step
+        def loss_fn(y, target):
+            return y.sum()
+
+        # Get stage module i from the entire model
+        def get_stage_module(stage_idx, num_stages):
+            # divide the model (8 layers) by the number of stages
+            layers_per_stage = total_layers // num_stages
+            assert layers_per_stage * num_stages == total_layers
+            # return offset so validation code can match partial layer back to orig model
+            offset = stage_idx * layers_per_stage
+            partial_model = nn.Sequential(
+                *full_model[offset : (stage_idx + 1) * layers_per_stage]
+            )
+            partial_model.to(self.device)
+            return partial_model, offset
+
+        # Apply DP to stage module
+        def apply_dp(partial_model, dp_type):
+            if dp_type == "FSDP":
+                # apply FSDP
+                mp_policy = MixedPrecisionPolicy(
+                    # TODO(whc) need to fix PP + FSDP-mixed-precision
```

Comment:
nit: It might be worth noting that if you do choose to make FSDP use bf16 mixed precision, then the unit test's `ref_model` needs to be changed.

Perhaps, it would be simpler for this unit test to just exclude `mp_policy` at all (in which case we just default to fp32 everything).

Reply:
I guess the code snippet is copied from titan. But sure, makes sense!